primitives/bridge/src/lib.rs

```diff
@@ -0,0 +1,406 @@
+        ensure!(
+            payload.len() < T::MaxMessagePayloadSize::get() as usize,
+            SendError::MessageTooLarge
+        );
+
+        // Ensure there is a registered channel we can transmit this message on
+        ensure!(
+            T::Channels::contains(&message.channel_id),
+            SendError::InvalidChannel
+        );
+
+        // Generate a unique message id unless one is provided
+        let message_id: H256 = message
+            .id
+            .unwrap_or_else(|| unique((message.channel_id, &message.command)).into());
+
+        // Fee not used
+        /*
+        let gas_used_at_most = T::GasMeter::maximum_gas_used_at_most(&message.command);
+        let fee = Self::calculate_fee(gas_used_at_most, T::PricingParameters::get());
+         */
+
+        let queued_message: VersionedQueuedMessage = QueuedMessage {
+            id: message_id,
+            channel_id: message.channel_id,
+            command: message.command.clone(),
+        }
+        .into();
+        // The whole message should not be too large
+        let encoded = queued_message
+            .encode()
+            .try_into()
+            .map_err(|_| SendError::MessageTooLarge)?;
+
+        let ticket = Ticket {
+            message_id,
+            channel_id: message.channel_id,
+            message: encoded,
+        };
+        let fee = Fee {
+            local: Default::default(),
+            remote: Default::default(),
+        };
+
+        Ok((ticket, fee))
+    }
+}
+
+impl ValidateMessage for () {
+    type Ticket = ();
+
+    fn validate(_message: &Message) -> Result<(Self::Ticket, Fee<u64>), SendError> {
+        Ok((
+            (),
+            Fee {
+                local: 1,
+                remote: 1,
+            },
+        ))
+    }
+}
+
+/// Message which is awaiting processing in the MessageQueue pallet
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq))]
+pub struct QueuedMessage {
+    /// Message ID
+    pub id: H256,
+    /// Channel ID
+    pub channel_id: ChannelId,
+    /// Command to execute in the Gateway contract
+    pub command: Command,
+}
+
+/// Enqueued outbound messages need to be versioned to prevent data corruption
+/// or loss after forkless runtime upgrades
+#[derive(Encode, Decode, TypeInfo, Clone, RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(PartialEq))]
+pub enum VersionedQueuedMessage {
+    V1(QueuedMessage),
+}
+
+impl From<QueuedMessage> for VersionedQueuedMessage {
+    fn from(x: QueuedMessage) -> Self {
+        VersionedQueuedMessage::V1(x)
+    }
+}
+
+impl From<VersionedQueuedMessage> for QueuedMessage {
+    fn from(x: VersionedQueuedMessage) -> Self {
+        match x {
+            VersionedQueuedMessage::V1(x) => x,
+        }
+    }
+}
+
+pub trait DeliverMessage {
+    type Ticket;
+
+    fn deliver(ticket: Self::Ticket) -> Result<H256, SendError>;
+}
+
+mod custom_do_process_message {
```

Comment:
Mainly because it adds some new imports and I wasn't sure if this is needed, so it would have been easier to remove it if it's all inside a module. I can move it to a separate file.

Reply:
fixed